src/FilesystemSyncTaskHelper.php

```diff
@@ -0,0 +1,349 @@
+        );
+    }
+
+    private function removeDirectories()
+    {
+        foreach (array_reverse($this->dirsToRemove) as $dir) {
+            if (count(glob("$dir/*")) === 0) {
+                rmdir($dir);
+            }
+        }
+    }
+
+    /**
+     * Whether the provided file should be skipped for syncing or not.
+     * @param array $file
+     * @return bool
+     */
+    private function skipFile($file)
+    {
+        // skip files starting with:
+        $skipsStartingWith = [
+            '_',
+            '.'
+        ];
+        foreach ($skipsStartingWith as $skipStartingWith) {
+            if (substr(basename($file['path']), 0, strlen($skipStartingWith)) === $skipStartingWith) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Whether the provided file should be synced with the database or not.
+     * @param array $fileTuple
+     * @return bool
+     */
+    private function canSyncFileToDatabase(array $fileTuple)
+    {
+        $realPath = $this->publicAssetAdapter->relativeToRealPath($fileTuple['path']);
+
+        if ($fileTuple['type'] == 'dir') {
+            // check if folder with hash exists in same folder
+            $sha1Short = basename($fileTuple['path']);
+            $folder = Folder::find($sha1Short);
+
+            $table = DataObject::singleton(File::class)->baseTable();
+            $fileExists = DB::prepared_query("SELECT COUNT(*) FROM \"$table\" WHERE \"FileHash\" LIKE ?", ["$sha1Short%"])->value();
+            if ($folder || $fileExists) {
+                $this->results['filesSkippedFromFilesystem']++;
+                return false;
+            }
+
+            // check if folder exists already in database
+            $path = $fileTuple['path'];
+            /** @var File $file */
+            $file = File::find($path);
+            if ($file) {
+                $this->results['filesSkippedFromFilesystem']++;
+                return false;
+            }
+        } else {
+            // all variants are removed
+            $parts = $this->flysystemAssetStore->parseFileID($fileTuple['path']);
+            if ($parts['Variant']) {
+                return false;
+            } elseif ($parts['Filename']) {
+                // hashed files are synced if their db entry is missing
+                $file = $this->flysystemAssetStore::getFileByFilename($fileTuple['path']);
+                if ($file) {
+                    $this->results['filesSkippedFromFilesystem']++;
+                    return false;
+                }
+
+                return true;
+            }
+
+            $file = File::find($fileTuple['path']);
+            if ($file) {
+                $this->results['filesSkippedFromFilesystem']++;
+                return false;
+            } elseif ($this->delete_broken) {
+                // file exists in filesystem but not DB and is in a hash folder
+                $contents = file_get_contents($realPath);
```

Comment:
Any specific reason why we're accessing the file directly here, rather than using flysystem?